arch/x86/adler32_avx.c

```diff
@@ -77,29 +81,70 @@ Z_INTERNAL uint32_t adler32_avx2(uint32_t adler, const unsigned char *buf, size_
+        * when done the scalar way, in GPRs:
+        
+        adler = (s1_unpack[0] % BASE) + (s1_unpack[1] % BASE) + (s1_unpack[2] % BASE) + (s1_unpack[3] % BASE) +
+                (s1_unpack[4] % BASE) + (s1_unpack[5] % BASE) + (s1_unpack[6] % BASE) + (s1_unpack[7] % BASE);
+
+        mov    $0x80078071,%edi // move magic constant into 32 bit register %edi
+        ...
+        vmovd  %xmm1,%esi // move vector lane 0 to 32 bit register %esi
+        mov    %rsi,%rax  // zero-extend this value to 64 bit precision in %rax
+        imul   %rdi,%rsi // do a signed multiplication with magic constant and vector element 
+        shr    $0x2f,%rsi // shift right by 47
+        imul   $0xfff1,%esi,%esi // do a signed multiplication with value truncated to 32 bits with 0xfff1 
+        sub    %esi,%eax // subtract lower 32 bits of original vector value from modified one above
+        ...
+        // repeats for each element with vpextract instructions
+
+        This is tricky with AVX2 for a number of reasons:
+            1.) There's no 64 bit multiplication instruction, but there is a sequence to get there
+            2.) There's ways to extend vectors to 64 bit precision, but no simple way to truncate
+                back down to 32 bit precision later (there is in AVX512) 
+            3.) Full width integer multiplications aren't cheap
+
+        We can, however, cast up to 64 bit precision on all 8 integers at once, and do a relatively
+        cheap sequence for horizontal sums. Then, we simply do the integer modulus on the resulting
+        64 bit GPR, on a scalar value
+        */
+
+ 
+        /* Will translate to nops */
+        __m128i s1lo = _mm256_castsi256_si128(vs1);
+        __m128i s2lo = _mm256_castsi256_si128(vs2);
+
+        /* Requires vextracti128 */
+        __m128i s1hi = _mm256_extracti128_si256(vs1, 1);
+        __m128i s2hi = _mm256_extracti128_si256(vs2, 1);
+        
+        /* Convert up to 64 bit precision to prevent overflow */
+        __m256i s1lo256 = _mm256_cvtepi32_epi64(s1lo);
+        __m256i s1hi256 = _mm256_cvtepi32_epi64(s1hi);
+        __m256i s2lo256 = _mm256_cvtepi32_epi64(s2lo);
+        __m256i s2hi256 = _mm256_cvtepi32_epi64(s2hi);
+
+        /* Sum vectors in existing lanes */
+        __m256i s1_sum = _mm256_add_epi64(s1lo256, s1hi256);
+        __m256i s2_sum = _mm256_add_epi64(s2lo256, s2hi256);
+        
+        /* In AVX2-land, this trip through GPRs will probably be unvoidable, as there's no cheap and easy
+         * conversion from 64 bit integer to 32 bit. This casting to 32 bit is cheap through GPRs 
+         * (just register aliasing), and safe, as our base is significantly smaller than UINT32_MAX */
+        adler = (uint32_t)(hsum(s1_sum) % BASE);
+        sum2 = (uint32_t)(hsum(s2_sum) % BASE);
+
+        vs1 = _mm256_set1_epi32(adler);
+        vs1 = _mm256_and_si256(vs1, vs_mask);
+
+        vs2 = _mm256_set1_epi32(sum2);
+        vs2 = _mm256_and_si256(vs2, vs_mask);
```

Comment:
See comment above... These lines are short, might be worth reducing them to 2 lines.